[NVPTX] Add TMA Bulk Copy Intrinsics

durga4github opened the pull request:
This patch adds a new variant of TMA Bulk Copy
intrinsics introduced in sm100+. This variant
has an additional byte_mask to select the bytes
for the copy operation.

* Selection is all done through table-gen now.
  So, this patch removes the corresponding
  SelectCpAsyncBulkS2G() function.
* lit tests are verified with a cuda-12.8 ptxas
  executable.

PTX Spec link:
https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-bulk-copy

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5323,6 +5323,20 @@ def int_nvvm_cp_async_bulk_shared_cta_to_global
+       llvm_i1_ty],        // Flag for cache_hint
+      [IntrConvergent, IntrArgMemOnly,
+       WriteOnly<ArgIndex<0>>, ReadOnly<ArgIndex<1>>,
+       NoCapture<ArgIndex<0>>, NoCapture<ArgIndex<1>>,
```
Since an async operation will presumably lead in reads/writes to these pointers after the intrinsic has returned isn't it incorrect to add nocapture here?

durga4github replied:
Yes, I think you are right. will update these.

durga4github replied:
Fixed in the latest revision,

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -2720,28 +2720,46 @@ void NVPTXDAGToDAGISel::SelectCpAsyncBulkTensorReduceCommon(SDNode *N,
 }
 
-void NVPTXDAGToDAGISel::SelectCpAsyncBulkS2G(SDNode *N) {
+void NVPTXDAGToDAGISel::SelectCpAsyncBulkS2GCommon(SDNode *N, bool HasMask) {
```
This function looks simple enough that it could be a table-gen pattern. Is there something I'm missing that prevents this?

durga4github replied:
This has a flag for the cache-hint. The flag itself gets dropped here, and we use the actual value when it is true (and ignore the value when it is false).

AlexMaclean replied:
Hmm, that's annoying that table-gen doesn't seem to allow you to drop/ignore an operand. It looks like you could to something like `(int_nvvm_cp_async_bulk_shared_cta_to_global i64:$dst, addr:$src, i32:$size, i64:$ch, -1)` for the case where we want the cache hint but I can't seem to figure out if there is a way to just ignore the cache-hint operand when the final operand is 0. @Artem-B do you know?

durga4github replied:
In general, the cache-hint pattern is common across a few families of TMA intrinsics. So, if there is an option to do this in table-gen, it can simplify a good chunk of the Cpp implementations.

Let us wait to hear from Artem,

durga4github replied:
In the latest revision, I have updated the lowering to be handled entirely through table-gen.
Please take a look and let me know what you think.

Artem-B replied:
> I can't seem to figure out if there is a way to just ignore the cache-hint operand when the final operand is 0.

You should be able to match a call with a constant argument. I think you can  do it literally.
https://github.com/llvm/llvm-project/blob/ae6e1276233ca541fdb2be1dde3074eb78277859/llvm/lib/Target/NVPTX/NVPTXIntrinsics.td#L1672
https://github.com/llvm/llvm-project/blob/ae6e1276233ca541fdb2be1dde3074eb78277859/llvm/lib/Target/NVPTX/NVPTXIntrinsics.td#L964

In the worst case you could use `PatLeaf` like we do for matching FP : https://github.com/llvm/llvm-project/blob/ae6e1276233ca541fdb2be1dde3074eb78277859/llvm/lib/Target/NVPTX/NVPTXIntrinsics.td#L14

durga4github replied:
Yes, I referred to the BFIND lowering you mentioned above, and we are matching the flag with a literal 0.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -525,18 +528,26 @@ class CpAsyncBulkStr<bit mc, bit ch> {
+
+def CP_ASYNC_BULK_S2G_CH : NVPTXInst<(outs),
+            (ins Int64Regs:$dst, ADDR:$src, Int32Regs:$size, Int64Regs:$ch),
+            !strconcat(CpAsyncBulkStr<0, 1>.S2G, " [$dst], [$src], $size, $ch;"), []>,
```
Nit: I personally prefer `#` to `!strconcat` in simple cases like this.

durga4github replied:
Done. Updated in the latest revision.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -525,18 +528,26 @@ class CpAsyncBulkStr<bit mc, bit ch> {
-defm CP_ASYNC_BULK_S2G_SHARED32 : CP_ASYNC_BULK_S2G<Int32Regs>;
+
+def CP_ASYNC_BULK_S2G_CH : NVPTXInst<(outs),
+            (ins Int64Regs:$dst, ADDR:$src, Int32Regs:$size, Int64Regs:$ch),
```
Should `$dst` be an `ADDR` as well?

durga4github replied:
Yes, I think it can be, but I will check.
I updated the src since it simplifies one level of opcode selection based on whether shared memory is 32 or 64 bits.

durga4github replied:
Updated `dst` also to be based on ADDR selection.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -516,6 +516,9 @@ class CpAsyncBulkStr<bit mc, bit ch> {
   string S2G = "cp.async.bulk.global.shared::cta.bulk_group"
                # !if(ch, ".L2::cache_hint", "");
 
+  // Shared to Global memory with bytemask
+  string S2G_BM = S2G # ".cp_mask";
+
```
Why add this here instead of just appending the suffix when using S2G below. I think adding this level of indirection between the instruction name and it's actual definition is making things a bit more confusing.

durga4github replied:
yes, the mask suffix here is rather short ;-)
Otherwise, we generally keep these strings in a class since the final instr name tends to be rather long.

I feel using a bool for the mask and generating one S2G will be more concise and consistent with other uses here. I will try this out.

durga4github replied:
Removed the S2G_BM variable and handling it just with S2G in the latest revision.

Artem-B commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -616,6 +616,7 @@ Syntax:
 .. code-block:: llvm
 
   declare void @llvm.nvvm.cp.async.bulk.shared.cta.to.global(ptr addrspace(1) %dst, ptr addrspace(3) %src, i32 %size, i64 %ch, i1 %flag_ch)
+  declare void @llvm.nvvm.cp.async.bulk.shared.cta.to.global.bytemask(..., i32 %size, i16 %mask, i64 %ch, i1 %flag_ch)
```
I'd make the mask the last argument to match the instruction argument order, though the `flag_ch` makes it impossible to match the instruction args exactly, but at least it would be a bit less surprising.

I wish we had a convenient way to have overloaded intrinsics with different number of arguments, so we don't need an explicit flag. Oh, well.

durga4github replied:
Done. I have moved the mask argument to the end in the latest revision.

I wanted to keep the flags (at the end) after the actual args, so I had kept it before the cache-hint. But I agree that retaining the instruction ordering is very helpful.

durga4github committed:
[NVPTX] Add TMA Bulk Copy Intrinsics

This patch adds a new variant of TMA Bulk Copy
intrinsics introduced in sm100+. This variant
has an additional byte_mask to select the bytes
for the copy operation.

* Selection is all done through tablegen now.
  So, this patch removes the corresponding
  SelectCpAsyncBulkS2G() function.
* lit tests are verified with a cuda-12.8 ptxas
  executable.

Signed-off-by: Durgadoss R <durgadossr@nvidia.com>

durga4github commented:
@Artem-B / @AlexMaclean , Ping for review on the latest revision..

AlexMaclean reviewed (approved):
LGTM

durga4github commented:
Rebased on top of the latest,
